spinoso-time/src/time/tzrs/math.rs

```diff
@@ -0,0 +1,491 @@
+use core::ops::{Add, Sub};
+use core::time::Duration;
+
+use tz::datetime::DateTime;
+
+use crate::time::tzrs::Time;
+use crate::NANOS_IN_SECOND;
+
+impl Time {
+    /// Rounds sub seconds to a given precision in decimal digits (0 digits by default). It returns
+    /// a new Time object. `ndigits` should be zero or a positive integer.
+    ///
+    /// Can be used to implement [`Time#round`]
+    ///
+    /// # Examples
+    /// ```
+    /// use spinoso_time::tzrs::Time;
+    /// let now = Time::local(2010, 3, 30, 5, 43, 25, 123456789);
+    /// let rounded = now.round(5);
+    /// assert_eq!(now.utc_offset(), rounded.utc_offset());
+    /// assert_eq!(123460000, rounded.nanoseconds());
+    /// ```
+    ///
+    /// [`Time#round`]: https://ruby-doc.org/core-2.6.3/Time.html#method-i-round
+    #[allow(clippy::cast_possible_truncation)]
+    #[allow(clippy::cast_sign_loss)]
+    #[inline]
+    pub fn round(&self, digits: u32) -> Self {
+        match digits {
+            9..=u32::MAX => *self,
+            digits => {
+                let local_time_type = *self.inner.local_time_type();
+                let mut unix_time = self.to_int();
+                let nanos = f64::from(self.nanoseconds()) / f64::from(NANOS_IN_SECOND);
+
+                let exponent_shift = f64::from(10_u32.pow(digits));
+
+                let rounded_nanos = (nanos * exponent_shift).round() / exponent_shift;
+                let mut new_nanos = (rounded_nanos * f64::from(NANOS_IN_SECOND)) as u32;
+                if new_nanos >= NANOS_IN_SECOND {
+                    unix_time += 1;
+                    new_nanos -= NANOS_IN_SECOND;
+                }
+
+                let dt = DateTime::from_timespec_and_local(unix_time, new_nanos, local_time_type)
+                    .expect("Could not round the datetime");
+                Self {
+                    inner: dt,
+                    offset: self.offset,
+                }
+            }
+        }
+    }
+}
+
+impl Add<Duration> for Time {
+    type Output = Self;
+
+    fn add(self, duration: Duration) -> Self::Output {
+        let unix_time = self.inner.unix_time();
+        let nanoseconds = self.inner.nanoseconds();
+        let offset = self.offset;
+
+        let duration_seconds = i64::try_from(duration.as_secs()).expect("Duration too large");
+        let duration_subsecs = duration.subsec_nanos();
+
+        let mut seconds = unix_time.checked_add(duration_seconds).expect("Duration too large");
+        let mut nanoseconds = nanoseconds.checked_add(duration_subsecs).expect("Duration too large");
```

Comment:
I think these `expect`s are similar to the ones in the `chrono` `ops::Add` impl. I kinda regret adding these since:

- they make fallibility hard (which we want so we can turn failures into at least a `RuntimeError`)
- I'm not super compelled to optimize for the ergonomics of `+` and `-` over explicit method calls since we control the only consumer of this code.

These are fine to leave in here for now, but I think I noticed you removed use of the `ops` overloads in the trampoline. In the long term, I'd like to remove the `math` impls from the `chrono` and `tzrs` impls.

Reply:
I removed the`ops` overloads because they weren't working :grimacing: haha (I was actually going to ask you about that). I've added them as a TODO in the Errors PR though since the encapsulation will likely be a `TimeError` etc.